src/front/wgsl/mod.rs

```diff
@@ -2034,6 +2267,106 @@ impl Parser {
         }))
     }
 
+    fn parse_constructor_type<'a>(
+        &mut self,
+        lexer: &mut Lexer<'a>,
+        word: &'a str,
+        type_arena: &mut UniqueArena<crate::Type>,
+        const_arena: &mut Arena<crate::Constant>,
+    ) -> Result<Option<ConstructorType>, Error<'a>> {
+        if let Some((kind, width)) = conv::get_scalar_type(word) {
+            return Ok(Some(ConstructorType::Scalar { kind, width }));
+        }
+
+        let partial = match word {
+            "vec2" => ConstructorType::PartialVector {
```

Comment:
it looks like we are not using the `enum ConstructorType` properly here. The `Partial*` variants are only appearing internally here. Perhaps, we could instead just have an "internal" enum for partial stuff and the main `enum ConstructorType` having no partial stuff? I suppose there wouldn't be needed `enum ConstructorType` at all, potentially, since we could just use `TypeInner`.

Reply:
It will only parse the component type if present. If not, will still return the partial variants. See the [last arm](https://github.com/teoxoy/naga/blob/6ae0745cccc45204c4673166842067755066db13/src/front/wgsl/mod.rs#L2366) of the match.

The partial variants are later used in the `parse_construction` function.